File: day8/day8.py
#!/usr/bin/env python3

import sys

BLACK = 0
WHITE = 1
TRANS = 2

def layers(seq, width, height):
    layers = []
    i = 0
    while i < len(seq):
        layer = []
        for y in range(height):
            row = []
            for x in range(width):
                row.append(int(seq[i]))
                i += 1
            layer.append(row)
        layers.append(layer)
    return layers

def count_digits(layer, digit):
    count = 0
    for row in layer:
        for x in row:
            if x == digit:
                count += 1
    return count

def minimal_layer(layers, digit):
    height, width = len(layers[0]), len(layers[0][0])
    mind = height * width
    minl = -1
    for i, l in enumerate(layers):
        c = count_digits(l, digit)
        if mind > c:
            mind = c
            minl = i

    if minl != -1:
        return layers[minl]

    return None

def stack_layers(layers):
    height, width = len(layers[0]), len(layers[0][0])
    output = [[TRANS for x in range(width)] for y in range(height)]
    for l in layers:
        for y in range(height):
            for x in range(width):
                if output[y][x] == TRANS:
                    output[y][x] = l[y][x]
    return output

def print_layer(layer):
    for row in layer:
        for x in row:
            if x == BLACK:
                sys.stdout.write(" ")
            elif x == WHITE:
                sys.stdout.write("#")
            else:
                sys.stdout.write(".")
        sys.stdout.write("\n")

with open("input", "rt") as file:
    seq = file.read().split("\n")[0]

l = layers(seq, 25, 6)
m = minimal_layer(l, BLACK)
if m:
    print("part1:", count_digits(m, WHITE) * count_digits(m, TRANS))

print("part2:")
print_layer(stack_layers(l))

File: day8/day8.c
#include <assert.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum
{
	BLACK = '0',
	WHITE = '1',
	TRANS = '2',

	WIDTH = 25,
	HEIGHT = 6,
};

static unsigned layer_count(size_t s)
{
	return s / WIDTH / HEIGHT;
}

static unsigned count_digits(const char *seq, unsigned layer, int digit)
{
	unsigned count = 0;
	for (unsigned j = layer * WIDTH * HEIGHT, n = j + WIDTH * HEIGHT;
	     j < n;
	     j++)
	{
		if (seq[j] == digit)
		{
			count++;
		}
	}
	return count;
}

static unsigned min_layer(const char *seq, size_t s, int digit)
{
	unsigned minc = WIDTH * HEIGHT;
	unsigned minj = UINT_MAX;
	for (unsigned j = 0, n = layer_count(s); j < n; j++)
	{
		unsigned c = count_digits(seq, j, digit);
		if (minc > c)
		{
			minc = c;
			minj = j;
		}
	}
	return minj;
}

static void flatten(char *seq, size_t s)
{
	char *dst = seq;
	char *src = seq + WIDTH * HEIGHT;
	s -= WIDTH * HEIGHT;
	while (s-->0)
	{
		if (dst == seq + WIDTH * HEIGHT)
		{
			dst = seq;
		}
		if (*dst == TRANS)
		{
			*dst = *src;
		}
		dst++;
		src++;
	}
}

static void print(const char *seq)
{
	for (int y = 0; y < HEIGHT; y++)
	{
		for (int x = 0; x < WIDTH; x++)
		{
			switch(*seq++)
			{
			case BLACK: printf(" "); break;
			case WHITE: printf("#"); break;
			default: printf("."); break;
			}
		}
		printf("\n");
	}
}


int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <input>\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s for reading\n", argv[1]);
		return -1;
	}

	char *buf = NULL;
	size_t bsize = 0;
	while (getline(&buf, &bsize, input) != -1)
	{
	}
	fclose(input);

	size_t size = strlen(buf);
	unsigned minj = min_layer(buf, size, BLACK);
	if (minj != UINT_MAX)
	{
		printf("part1: %u\n",
		       count_digits(buf, minj, WHITE) *
		       count_digits(buf, minj, TRANS));
	}

	flatten(buf, size);
	printf("part2:\n");
	print(buf);

	free(buf);
	return 0;
}
